database/mysql/mysql.go

```diff
@@ -82,21 +84,31 @@ func WithInstance(instance *sql.DB, config *Config) (database.Driver, error) {
 		config.MigrationsTable = DefaultMigrationsTable
 	}
 
-	conn, err := instance.Conn(context.Background())
-	if err != nil {
+	if err := mx.ensureVersionTable(); err != nil {
 		return nil, err
 	}
 
-	mx := &Mysql{
-		conn:   conn,
-		db:     instance,
-		config: config,
+	return mx, nil
+}
+
+// instance must have `multiStatements` set to true
+func WithInstance(instance *sql.DB, config *Config) (database.Driver, error) {
+	if err := instance.Ping(); err != nil {
```

Comment:
Move this to `WithConnection()` and use `conn.PingContext()`. `DB.Ping()` will [create a connection to ping](https://golang.org/pkg/database/sql/#DB.Ping) anyways

Reply:
The problem I am trying to solve is that this class does does unexpected stuff to my connection pool.
If I inject an active pool with `n` idle connections to this class, do my stuff and close it, I expect my pool to be returned to it's initial state, which means an active pool with `n` idle connections.
But currently I can only choose between ending-up with a closed pool (because `Close` not only returns the connection to the pool, but also closes the pool itself), or (if I don't call `Close`) a pool with `n-1` idle connections and `1` allocated connection that will be stuck forever

So the last thing I want is for `WithConnection` to create new connections in a pool that I did not even inject in the first place.

If the connection I inject is not working, I expect the process to fail without side-effects on my pool.

If the connection is not working, doing `conn.PingContext().DB.Ping()` would create a new connection in the pool and ping it, which does not fix the injected connection anyway.

Reply:
My understanding is that connections are pooled at the `sql.DB` instance level so `conn.PingContext()` [shouldn't create a new connection](https://github.com/golang/go/blob/go1.16.5/src/database/sql/sql.go#L1869) e.g. connections are won't create new connections

Reply:
Sorry, I was misunderstanding your question.
I added the context to `WithConnection`, as well as calling `PingContext`.
I cannot add it as a parameter in `WithInstance` without breaking change, so it should be done along with #14.